src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs

```diff
@@ -436,16 +436,35 @@ public Task<List<StreamSubscription>> GetAllSubscriptions(StreamId streamId, ISt
 
         private async Task NotifyProducersOfRemovedSubscription(GuidId subscriptionId, StreamId streamId)
         {
-            int numProducers = State.Producers.Count;
-            if (numProducers > 0)
+            int numProducersBeforeNotify = State.Producers.Count;
+            if (numProducersBeforeNotify > 0)
             {
-                if (logger.IsVerbose) logger.Verbose("Notifying {0} existing producers about unregistered consumer.", numProducers);
+                if (logger.IsVerbose) logger.Verbose("Notifying {0} existing producers about unregistered consumer.", numProducersBeforeNotify);
 
                 // Notify producers about unregistered consumer.
                 List<Task> tasks = State.Producers.Where(producerState => IsActiveProducer(producerState.Producer))
-                                                  .Select(producerState => producerState.Producer.RemoveSubscriber(subscriptionId, streamId))
+                                                  .Select(producerState => NotifyProducerOfRemovedSubscriber(producerState, subscriptionId, streamId))
                                                   .ToList();
                 await Task.WhenAll(tasks);
+                //if producers got removed
+                if (State.Producers.Count < numProducersBeforeNotify)
+                    await this.WriteStateAsync();
+            }
+        }
+
+        private async Task NotifyProducerOfRemovedSubscriber(PubSubPublisherState producer, GuidId subscriptionId, StreamId streamId)
+        {
+            try
+            {
+                await producer.Producer.RemoveSubscriber(subscriptionId, streamId);
+            }
+            catch (GrainExtensionNotInstalledException)
+            {
+                RemoveProducer(producer);
+            }
+            catch (ClientNotAvailableException)
```

Comment:
This is the correct behavior in this case.
This will be thrown when an Orleans client is producing events on a stream and loses connection to the cluster.
This logic was added to the add consumer, but not to the remove consumer when it should apply to both.